### PR TITLE
Fixes decoding twice in validation

### LINKED ISSUE #70: Decoding twice in validation step
See here: https://github.com/CUNY-CL/yoyodyne/blob/master/yoyodyne/models/base.py#L219

Previously, I did this so I could first predict with greedy search to get an accuracy, and then predict with teacher forcing to compute a loss taht is comparable to how the train loss is computed. Currently, we predict twice with whatever is in the batch from the validation dataloader. IIRC this should have the gold targets in it.

This means that we are making identical validation predictions twice, neither of which is greedy -- they both have access to the gold history and use teacher forcing. I think we probably want to compute validation accuracy without the gold history, for which we would need to pass the batch through with no targets on it.

Then, we can either keep predicting a 2nd time to compute a loss with teacher forcing, or just compute the loss directly from the greedy predictions (and thus only decode once during validation).

@kylebgorman thoughts on this?

### PR DESCRIPTION
**Pull request by Adamits**
- Updates validation step to only call model forawrd once.
- Adds teacher_forcing param to  forward and decode methods.
- Calls validation step with teacher_forcing=False.
- Updates  methods to decode until EOS AND at least length of target in teacher_forcing mode.
- Calls predict step with teacher_forcing=False.

Closes #70 

Adds `teacher_forcing`, which tells the model to decode using the gold history, rather than the predicted history. This is typically how encoder-decoder inflection models are trained, but during validation/prediction, we force teacher_forcing to be False.

I additionally implemented a check that makes validation more efficient: we supply the target tensor, but only use it for early stopping on decoding if the decoder has predicted an EOS token for every sequence in the batch. This ensures that we can decode greedily up to at least the length of the targets, and then compute a loss in which we need `len(predicted == len(targets)`.

*IMPORTANT* 
- For Transducer, teacher_forcing means we get the gold actions from the expert. @bonham79 does this make sense? Or should I ignore this new parameter entirely in that architecture?
- Transducer errors with:

```
  File "/Users/adamwiemerslage/nlp-projects/morphology/yoyodyne/yoyodyne/models/expert.py", line 105, in lookup
    return self.w2i[word]
KeyError: ConditionalIns(new=0)
```

I reverted all changes to test, and still reproduce the error. This is on the ConLL-SIGMORPHON 2017 inflection shared task English medium setting data, with the dev set. I suspect this is related to an old issue that happened when there was an OOV in the eval set. @bonham79 -- do you know if that issue still exists in the transducer?

### CODE REVIEW

**Comment by tbartley94** [comment]
>For Transducer, teacher_forcing means we get the gold actions from the expert. @bonham79 does this make sense? Or should I ignore this new parameter entirely in that architecture?

I never used that flag with the Transducer so it should not have an effect. The teacher forcing for that uses a different query from the SED parameters. 

>  I suspect this is related to an old issue that happened when there was an OOV in the eval set. @bonham79 -- do you know if that issue still exists in the transducer?

The OOV issue (iirc) was related to OOV _features_ , are you referring to that?

**Comment by Adamits** [comment]
> > For Transducer, teacher_forcing means we get the gold actions from the expert. @bonham79 does this make sense? Or should I ignore this new parameter entirely in that architecture?
> 
> Yes that makes sense. The query from the expert is equivalent to teacher forcing. The main difference is that instead of the gold being static, it's recalculated at each step of transcription to allow correction during bad predictions.
> 
> > I suspect this is related to an old issue that happened when there was an OOV in the eval set. @bonham79 -- do you know if that issue still exists in the transducer?
> 
> The OOV issue (iirc) was related to OOV _features_ , are you referring to that?

Ah, yes possibly. Do you remember if we fixed that? I can debug this a bit, but I don't think its related to my changes to would rather not make changes on this PR.

**Comment by kylebgorman** [comment]
Sorry y'all, I should be able to get to this tomorrow.

It does sound like the transducer is broken at head from your description---that should be solved separately once Travis is recovered from his ICASSP jaunt---happy to consult on that.

**Comment by tbartley94** [comment]
> Ah, yes possibly. Do you remember if we fixed that? I can debug this a bit, but I don't think its related to my changes to would rather not make changes on this PR.

iirc no since it's being addressed in feature encoder update. that should be done next week when i'm stateside.

**Comment by tbartley94** [comment]
> It does sound like the transducer is broken at head from your description---that should be solved separately once Travis is recovered from his ICASSP jaunt---happy to consult on that.

yeah, i'm trying to integrate the transducer more into basic lstm behavior so updates are easier. i give it about two more major updates and issues like these shouldnt' be popping up as often.

File: yoyodyne/models/pointer_generator.py
Comment (kylebgorman) [inline yoyodyne/models/pointer_generator.py:167]: ditto from above.

File: yoyodyne/models/pointer_generator.py
Comment (kylebgorman) [inline yoyodyne/models/pointer_generator.py:234]: ditto

File: yoyodyne/models/transformer.py
Comment (kylebgorman) [inline yoyodyne/models/transformer.py:253]: ditto

**Review by kylebgorman — COMMENTED** [review]
Thanks for this, this is a big update of sorts.

I have to ask whether there's some function we could write that would reduce the boilerplate...is there?

Other than my nits, I'd suggest you go ahead and submit even if you're getting an OOV error on some data in the transducer, so long as 1) you believe it to be independent and 2) there's a bug report/issue reported for that issue so we don't forget about it.

**Comment by Adamits** [comment]
> > It does sound like the transducer is broken at head from your description---that should be solved separately once Travis is recovered from his ICASSP jaunt---happy to consult on that.
> 
> yeah, i'm trying to integrate the transducer more into basic lstm behavior so updates are easier. i give it about two more major updates and issues like these shouldnt' be popping up as often.

Cool -- sorry for the false alarm! Enjoy ICCASP.

**Comment by Adamits** [comment]
> I have to ask whether there's some function we could write that would reduce the boilerplate...is there?

Ah good point. I think there is. In fact,  I think the `decode(...)` implementation for LSTM models is setup fairly differently than transformer and we could make these much more similar in a way that would reduce repeated code. iirc Travis had pointed this out in a PR or issue a while back.

I can work out those updates sometime before the week ends.

**Comment by tbartley94** [comment]
> > I have to ask whether there's some function we could write that would reduce the boilerplate...is there?
> 
> Ah good point. I think there is. In fact, I think the `decode(...)` implementation for LSTM models is setup fairly differently than transformer and we could make these much more similar in a way that would reduce repeated code. iirc Travis had pointed this out in a PR or issue a while back.
> 
> I can work out those updates sometime before the week ends.

Don't worry about the boilerplate for models, I have it cleaned up in an upcoming PR.

**Comment by Adamits** [comment]
Updated per the PR comments. Leaving updates around reduction of boilerplate and conforming `decode` methods to Travis' PR.

File: yoyodyne/models/base.py
Comment (kylebgorman) [inline yoyodyne/models/base.py:260]: I notice that `batch` and `predictions` aren't passed in. Is it supposed to work like that, getting them from context? Seems scary.

File: yoyodyne/models/lstm.py
Comment (kylebgorman) [inline yoyodyne/models/lstm.py:247]: Why can't this switch on whether or not `target` is null?

File: yoyodyne/models/lstm.py
Comment (kylebgorman) [inline yoyodyne/models/lstm.py:202]: Why can't this switch on whether or not `target` is null? The documentation for `target` seems to suggest that's what it does.

File: yoyodyne/models/lstm.py
Comment (kylebgorman) [inline yoyodyne/models/lstm.py:423]: Suggest making this:

`batch.target.padded if teacher_forcing else None`

and the removing the following argument, and adjusting interface to respect  that.

File: yoyodyne/models/pointer_generator.py
Comment (kylebgorman) [reply yoyodyne/models/pointer_generator.py:224]: Make this consistent with the rest of them, whatever you do here.

**Review by kylebgorman — COMMENTED** [review]
See my comments about on `validation_epoch_end` and how we're twice encoding whether or not we're using teacher forcing in the helpers.

File: yoyodyne/models/base.py
Comment (kylebgorman) [reply yoyodyne/models/base.py:260]: Also wouldn't you want to null out the whole target portion of the batch and not just the padded vector?

**Comment by bonham79** [comment]
> Ah, yes possibly. Do you remember if we fixed that? I can debug this a bit, but I don't think its related to my changes to would rather not make changes on this PR.

Checked the transducer, it's not really a bug but more because that dataset is poorly formatted, so there's character symbols in dev that don't occur in train. Due to the transducer predicting edits, not actual characters, there's no logic in place for predicting characters that have never be seen. I can fix it but for me that falls under the category of data processing.

Side note: That, uh, CONLL set had some interesting words in it...

**Comment by kylebgorman** [comment]
> Checked the transducer, it's not really a bug but more because that dataset s poorly formatted, so there's character symbols in dev that don't occur in train. 

So can we fix this dataset automatically? Can you include the dataset formatting bug report here?

> Due to the transducer predicting edits, not actual characters, there's no logic in place for predicting characters that have never be seen. I can fix it but for me that falls under the category of data processing.

I wonder if we can catch and convert this error to something more informative?

**Comment by bonham79** [comment]
>
> So can we fix this dataset automatically? Can you include the dataset formatting bug report here?

The fix would just be standard post processing (force single casing, stripping punctuation). Best I can provide for a bug report is just a comparison of dev vocabulary set vs train vocabulary set. 

It's now fixed in https://github.com/CUNY-CL/yoyodyne/pull/72 by inserting a <UNK> character when the desired output is outside training. 
 
> 
> I wonder if we can catch and convert this error to something more informative?

We could compare indexes from dev and train when doing an initial dataset instantiation just to flag if vocabularies aren't subsets. Issue is that I think this is just a transducer issue so a bit wasteful. Also, this strikes me as something that users should be responsible for in the data curation stage.

**Comment by kylebgorman** [comment]
> It's now fixed in #72 by inserting a character when the desired output is outside training

If that's so let's forget about it. You do that with the bit about the unk index, right?

> We could compare indexes from dev and train when doing an initial dataset instantiation just to flag if vocabularies aren't subsets. Issue is that I think this is just a transducer issue so a bit wasteful. Also, this strikes me as something that users should be responsible for in the data curation stage.

I agree the user should be responsible for this but the problem I see is that the error message is not informative and gives no hint to the user how to fix it. The solution should be simple and cheap: catch the `KeyError` (if that's what it is) and convert it to more specific error message. (I think that try/except code is very cheap at runtime.)

**Comment by bonham79** [comment]
> If that's so let's forget about it. You do that with the bit about the unk index, right?
> 
yep

> I agree the user should be responsible for this but the problem I see is that the error message is not informative and gives no hint to the user how to fix it. The solution should be simple and cheap: catch the `KeyError` (if that's what it is) and convert it to more specific error message. (I think that try/except code is very cheap at runtime.)

Error handling is strangely expensive for some reason. It's fine for instantiation but for forward passes it's a resource drain.  I think the fix with UNK makes it more consistent with other behavior so should be good.

**Comment by kylebgorman** [comment]
I approve this modulo the unaddressed issues from my earlier review. I think @Adamits is busy with something else for a few days though...

File: yoyodyne/models/base.py
Comment (Adamits) [reply yoyodyne/models/base.py:260]: > I notice that batch and predictions aren't passed in. Is it supposed to work like that, getting them from context? Seems scary.

I don't follow what you mean here.

> Also wouldn't you want to null out the whole target portion of the batch and not just the padded vector?

Ideally, I think yes. I can't remember which model behavior forced me to do this. Will retrace my steps and update (either with a fix, or clearer justification :D)

File: yoyodyne/models/lstm.py
Comment (Adamits) [reply yoyodyne/models/lstm.py:247]: See my email: we can still use the target for early stopping if provided, even when evaluating without teacher forcing. Is that too tricky/confusing?

File: yoyodyne/models/base.py
Comment (Adamits) [reply yoyodyne/models/base.py:260]: Not sure if this is what you mean here, but thinking about it more, it might be a bad idea to force remove targets here? On the one hand, I don't think `predict` should ever let the model see the targets. On the other hand, this should probably be handled elswhere that is less implicit.

The problem is that currently there is not obvious place in the code to say "don't use the target" (I guess setting teacher forcing to False handles this...). And I think its reasonable to pass in a test file with targets and expect the code to not use them at runtime when predicting.

Comment (kylebgorman) [reply yoyodyne/models/base.py:260]: I mean i don't understand why this doesn't complain that `batch` is undefined. The following snippet of Python should give a `NameError` on `baz` because `baz` hasn't been defined. Similarly, `batch` and `prediction` are undefined in this method. We don't rely on this kind of scoping anywhere else in the entire project to my knowledge: we pass in the arguments we refer to.

```python
def foo(bar):
   return baz + baz
```

Yes, we should pass this as a kwarg and not forcibly remove the target.

**Review by kylebgorman — COMMENTED** [review]
More notes.

File: yoyodyne/models/base.py
Comment (kylebgorman) [reply yoyodyne/models/base.py:260]: Ah that's right, sorry to waste your time.

**Comment by Adamits** [comment]
> Checked the transducer, it's not really a bug but more because that dataset is poorly formatted, so there's character symbols in dev that don't occur in train. Due to the transducer predicting edits, not actual characters, there's no logic in place for predicting characters that have never be seen. I can fix it but for me that falls under the category of data processing.

Is OOV poor formatting? I think of it as a valid test case.

> Side note: That, uh, CONLL set had some interesting words in it...

Yes i've noticed that too :D.

Looking further down the thread about this being on the user:

> I agree the user should be responsible for this but the problem I see is that the error message is not informative and gives no hint to the user how to fix it. The solution should be simple and cheap: catch the KeyError (if that's what it is) and convert it to more specific error message. (I think that try/except code is very cheap at runtime.)

Could we catch it when loading data? I think our code should load dev/test data, then load the train index. After loading the train index, we should be able to find all OOVs, and log a warning? And then do you think the user is supposed to preprocess their data and replace any OOV with some <UNK> symbol? We could alternatively supply an option in the library, not totally sure what is right. Actually come to think of it I thought we already had an UNK symbol in the special vocab that we use to encode stuff not in the index so maybe I am misunderstanding the problem...

File: yoyodyne/models/lstm.py
Comment (Adamits) [reply yoyodyne/models/lstm.py:423]: Sorry I am a bit all over the place today, but also just realized that a user *should* be able to specify `--target_col=0` to control this from the CLI. Idk that we ever actually tested that though.

Comment (kylebgorman) [reply yoyodyne/models/lstm.py:423]: Yes, that'd be great if we could just document that and have it work that way! (I'd also log the teacher/student status in the `train` or `predict` main.)

Comment (kylebgorman) [reply yoyodyne/models/lstm.py:423]: On reflection I don't think this will (or could) work for training, just prediction. In training, we need to read in the target whether or not we're using it during decoding. In prediction though, this might...

Comment (Adamits) [reply yoyodyne/models/lstm.py:423]: Right agreed. Maybe I need to think this through better from the perspective of what is most logical for a user.

I think the teacher forcing flag is useful for making it easy to understand the code during training (check the validation step code, see that we set teacher forcing to False during validation) -- and that this is a pretty natural behavior.

During prediction, we _could_ require a user to set the target col to 0, but then again, I think it is also a natural assumption that `predict_step` should not have access to the targets, even if they are in the file. Maybe the best course of action is to leave it up to the user, but log a warning if target is not zero'd out. 

My only concern is if this warning will be understandable :D. I think the default behavior would be that we predict with teacher_forcing=False, but that, as in validation, we can stop early once we have predicted at least as many tokens as the target length.

Comment (kylebgorman) [reply yoyodyne/models/lstm.py:423]: At the highest level I think the interface should be:

* `--forcing` defaults to "default" (teacher during training, student during prediction)
* `--forcing student` does what you think regardless of whether we're doing training or prediction
* `--forcing teacher` ... ditto

Regardless of how the dataset args are passed, you should need a valid target column for teacher forcing and it shouldn't matter for student forcing.

Comment (Adamits) [reply yoyodyne/models/lstm.py:423]: Been trying to add this as an argument that we call from train/predict, but I am unsure of how pass an argument through the training loop in lightning, and cannot find anything in docs/issues.

Now I am wondering if we should consider `teacher_forcing` as only valid during training, and add it as a model attribute (if we later extend it to a teacher forcing "ratio", that can still live on the model). Then, the question still remains, where should we tell the forward/decode method to ignore the gold symbols if they exist?

One clean solution (1) is to always force `teacher_forcing=False` in the validate/predict steps like we are doing now. Instead, I suppose we could (2) check if the model is in `eval()` mode? I think (1) makes the most sense. But I keep worrying if the semantics of this is confusing. 

Additional point: I would rather keep the argument as `--teacher_forcing` and have it as either a bool or float if we want a ratio. I think having forcing types is a bit weird since IDK if `student_forcing` is a common term, not if there could be any other possible type of forcing that is not expressed by an "amount" of teacher forcing.

**Comment by Adamits** [comment]
Ok, I have commited an update to the design.

`teacher_forcing` is now only relevant as a training argument, and is set on the model. I have overloaded it as a local variable to the `decode` method (plz let me know if this is problematic and maybe we can come up with a new name. Since this PR is really just about validation issues, maybe I can deal with tweaks to this in a separate issue?).

I think after Travis' large refactor, maybe I should change teacher_forcing to a `float`, and add a method that `decode` calls before each step to decide whether or not to pass the previous gold step or not. It will default to 1, and be set to 0 in eval mode -- future work though.

File: yoyodyne/models/lstm.py
Comment (bonham79) [reply yoyodyne/models/lstm.py:423]: Jumping in on this, unless it was changed a few prs ago, I believe I wrote into `predict` that `target-col` is automatically set to `0` just for these concerns. So during the predict stage, the upstream should be stopping any presence of target.

**Comment by bonham79** [comment]
Oh sorry, email was clogged and didn't see these comments.
 
> Is OOV poor formatting? I think of it as a valid test case.
> 

Mmmm, the `UNK` edit character allows management so it's not too hard  to do. But, given that our domain is string transduction, I don't see a good rationale for it. Why would you train your model on a dataset that doesn't contain the alphabet of your potential strings? You should be able to find a dataset with minimal coverage (outside of cryptography and potentially Hanji/Kanji, and even then...) 

And in the case you don't find that dataset...what do you hope to learn? Idk, it seems like too low a bar to accommodate for imo.

> 
> Looking further down the thread about this being on the user:
> 
> > I agree the user should be responsible for this but the problem I see is that the error message is not informative and gives no hint to the user how to fix it. The solution should be simple and cheap: catch the KeyError (if that's what it is) and convert it to more specific error message. (I think that try/except code is very cheap at runtime.)
> 
> Could we catch it when loading data? I think our code should load dev/test data, then load the train index. After loading the train index, we should be able to find all OOVs, and log a warning? And then do you think the user is supposed to preprocess their data and replace any OOV with some symbol? We could alternatively supply an option in the library, not totally sure what is right. Actually come to think of it I thought we already had an UNK symbol in the special vocab that we use to encode stuff not in the index so maybe I am misunderstanding the problem...

The issue was that for transducer there was an UNK symbol but not an UNK edit action. So it knew the symbol existed, it just didn't have a way to respond to it. Now it should have the option to choose an UNK symbol, so the error shouldn't pop up.

Re: OOV checking. I like that idea. Thinking it would work best as (yet anothe!) CLI flag. Probably like `--check-oov` that reads an index for the `dev` set and raises a warning. Would be a new PR though.

**Comment by kylebgorman** [comment]
I’m fine pushing OOV handling into another PR.

K

**Comment by bonham79** [comment]
> I think after Travis' large refactor, maybe I should change teacher_forcing to a `float`, and add a method that `decode` calls before each step to decide whether or not to pass the previous gold step or not. It will default to 1, and be set to 0 in eval mode -- future work though.

That's gonna be a bit of a headache in future updates. `teacher_forcing` reads like such a `bool` variable that having it serve as a `float` will cause some bugs. (Not nay-saying, just  acknowledging that I'm enough of an idiot that I will make that mistake.)

**Comment by kylebgorman** [comment]
I am now thinking we may have jumped the gun exposing it to the user, and would be better off just forcing teacher forcing during training and student forcing during inference. I think we should actually hyperparameter-experiment with a probabilistic mix of student forcing during training (i.e., try to see if that Nicolai paper is right for the kind of problems we care about) and only then it out as a flag if it makes an impact.

File: yoyodyne/models/lstm.py
Comment (kylebgorman) [inline yoyodyne/models/lstm.py:250]: I have a minor suggestion. Put an assertion (or module-specific error) here to catch when `teacher_forcing` is True but `target` is None. It shouldn't be reachable but if it is that'd be better than `TypeError: 'NoneType' object is not subscriptable`.

```python
if teacher_forcing:
   assert target is not None, "Teacher forcing requested but no target provided"
   ...
```

File: yoyodyne/models/lstm.py
Comment (kylebgorman) [reply yoyodyne/models/lstm.py:423]: I think what we have here is good for now...resolving.

File: yoyodyne/models/transformer.py
Comment (kylebgorman) [inline yoyodyne/models/transformer.py:207]: typo: `predicitons`

**Review by kylebgorman — APPROVED** [review]
Okay LGTM.

I suggest assertions to catch teacher forcing requested without target passed. 

I note that there is some inconsistency in interface:

* the LSTM and pointer generator accept explicit `teacher_forcing` as an argument to `decode`
* the transformer reads it from `self.teacher_forcing`
* the transducer does both, I think. 

Is there anything to be done about this?

**Comment by Adamits** [comment]
> Okay LGTM.
> 
> I suggest assertions to catch teacher forcing requested without target passed.

Ok, I will add an assertion good idea.

> I note that there is some inconsistency in interface:
> 
> * the LSTM and pointer generator accept explicit `teacher_forcing` as an argument to `decode`
> * the transformer reads it from `self.teacher_forcing`
> * the transducer does both, I think.
> 
> Is there anything to be done about this?

I believe transformer is the only one that is inconsistant. This is because the interface is already inconsistent, which, I think @bonham79  is changing anyway?

**Comment by kylebgorman** [comment]
> I believe transformer is the only one that is inconsistant. This is because the interface is already inconsistent, which, I think @bonham79 is changing anyway?

He confirms this.

**Comment by bonham79** [comment]
> > Okay LGTM.
> > I suggest assertions to catch teacher forcing requested without target passed.
> 
> Ok, I will add an assertion good idea.
> 
> > I note that there is some inconsistency in interface:
> > 
> > * the LSTM and pointer generator accept explicit `teacher_forcing` as an argument to `decode`
> > * the transformer reads it from `self.teacher_forcing`
> > * the transducer does both, I think.
> > 
> > Is there anything to be done about this?
> 
> I believe transformer is the only one that is inconsistant. This is because the interface is already inconsistent, which, I think @bonham79 is changing anyway?

Correct.

**Comment by tbartley94** [comment]
> > I believe transformer is the only one that is inconsistant. This is because the interface is already inconsistent, which, I think @bonham79 is changing anyway?
> 
> Correct.

To follow up, if we do this pr before the other one, it'll give me a chance to clean up the interface further for consistency. So assume interface changes will happen in follow up.

**Comment by tbartley94** [comment]
> FYI I **believe** need to make some fixes to the predict script (different PR) -- not sure if I should wait for your PR to merge first? I also have some work adding wandb sweeps, but that should not interfere too much with your work AFAIK.
 
IIRC there's no change to predict behavior (modules are loaded as part of the model) so you're good to go to make changes async to other pr.

**Comment by kylebgorman** [comment]
Okay should I merge?

**Comment by Adamits** [comment]
> Okay should I merge?

Yes!

**Comment by tbartley94** [comment]
> Okay should I merge?

yeeep

**Pull request merged**
